### ISSUE TITLE
object/embed

### ISSUE THREAD

**Issue by annevk** [slightlyoff/ServiceWorker#249]
Depending on the response, object/embed either create a browsing context or are embedding elements.

Service workers do not allow for this design as elements that create nested browsing contexts would always go through their own service worker.

Therefore these elements either become wormholes or will never be able to create a nested browsing context once you active a service worker.

This seems bad.

**Comment by jungkees** [comment]
I think we can close this issue for now.

**Comment by annevk** [comment]
Why? Don't we ever want to address this?

**Comment by jungkees** [comment]
I thought returning null for potential client requests in Handle Fetch makes sense, and we didn't come up with any further discussion for that. 

Let's think about that then. Re-opening.

**Comment by annevk** [comment]
We should perhaps introduce attributes in HTML that lock these elements to being either a browsing context or embedding context (upon insertion or some such) and then service workers can work with them.

**Comment by jakearchibald** [comment]
I think the type attr can do that?

**Comment by annevk** [comment]
For `object` I think that `typemustmatch` would also have to be set. But yeah, some set of rules around that would probably do the trick.

**Comment by patrickkettner** [comment]
To clarify - there is currently no way for a SW to serve a swf file that is defined as the `data` attribute on an `object`/`embed`, correct?

**Comment by annevk** [comment]
Correct.

**Comment by babatakao** [comment]
Because of this issue, it is hard to use SVG on SW. Since images referenced from SVG specified by <img> are not shown, it is needed to use `<object>`. (iframe or inline SVG may be the solution, but they cannot be always adopted.)

There are no activity for long time. Any plan to address it?

**Comment by jakearchibald** [comment]
Iframe is definitely the closest thing right now. Although `<img>` works if you don't need scripting, and as you say there's inline SVG.

Can you go into a bit more details in terms of why `iframe` doesn't work as a solution here?

**Comment by babatakao** [comment]
The root cause is for the existing contents. I assume an EPUB reading system. Quite a lot of EPUB files use object to load SVG.

**Comment by jakearchibald** [comment]
We can only solve this if we know in advance whether the `<object>` will create a browsing context or not. That means setting the `type`, and whatever ensures strict type matching (it used to be `typemustmatch` but that seems to be gone now, replaced with `X-Content-Type-Options: nosniff` perhaps?).

However, if you're saying that this problem only exists when existing contents cannot be changed, then there's no point us fixing it, as it would still depend on existing content changing.

**Comment by babatakao** [comment]
Thank you, I've understood the difficulty.

Just on the use case I assumed, most of `<object>` elements already have `type` attribute and it's easy to add "X-Content-Type-Options" header. So conditional support for object element is welcome.

That said, I will consider the way to modify contents with iframe.

**Comment by zcorpan** [comment]
It seems that `object` creates a browsing context for images, documents and SVGs in Chromium and WebKit. In Gecko, images do not create a browsing context. Also see https://github.com/whatwg/html/issues/7939

- `object` https://software.hixie.ch/utilities/js/live-dom-viewer/saved/13106
- `embed` https://software.hixie.ch/utilities/js/live-dom-viewer/saved/13107

Does that change the situation for this issue?

Also related is the YouTube `embed` rewrite issue, and whether that should create a browsing context, see https://github.com/whatwg/html/issues/2390#issuecomment-2363164672

**Comment by yoshisatoyanagisawa** [comment]
Bug Triage: infeasible
https://w3c.github.io/ServiceWorker/#on-fetch-request-algorithm
```
8. If request’s [destination](https://fetch.spec.whatwg.org/#concept-request-destination) is either "embed" or "object", then:
  1. return null.
```
These objects are not intercepted.

**Comment by zcorpan** [comment]
Why is it infeasible? If `object` and `embed` always create a browsing context (until they decide to show fallback), it seems like it should be possible to handle them similarly to `iframe`.

**Comment by yoshisatoyanagisawa** [comment]
You're right to ask why this is different from an `<iframe>`. The key challenge is that `<object>` and `<embed>` don't always create a new browsing context; it depends entirely on the response's MIME type, which is unknown when the request is made (which is what I have understood from the discussion).

An `<iframe>` is always a navigation request for a new context. In contrast, `<object>` and `<embed>` create an ambiguity: should the Service Worker treat the request as a simple subresource fetch for the current page, or as a navigation that should be handled by the new context's own Service Worker?

Resolving this ambiguity after the response has been received would require complex mechanisms, like restarting the request entirely. Given the significant implementation cost for what are now niche use cases, it has been difficult to justify prioritizing a complete solution.

Therefore, I support maintaining the current behavior of not intercepting these requests. It strikes the right balance between platform completeness and the practical cost of implementation.

**Comment by zcorpan** [comment]
Making `object` and `embed` always create a browsing context is being tracked in https://github.com/whatwg/html/issues/7939

This issue is a blocker for being able to test YouTube flash embed rewrite: https://github.com/whatwg/html/issues/2390#issuecomment-2348414523

**Comment by yoshisatoyanagisawa** [comment]
Thanks. I agree that aligning the behavior of `<object>` and `<embed>` with `<iframe>` is the right direction for consistency with Service Workers.
